src/System.Management.Automation/engine/Subsystem/Prediction/CommandPrediction.cs

```diff
@@ -0,0 +1,174 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Management.Automation.Language;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Management.Automation.Subsystem
+{
+    /// <summary>
+    /// The class represents the prediction result from a predictor.
+    /// </summary>
+    public sealed class PredictionResult
+    {
+        /// <summary>
+        /// Gets the Id of the predictor.
+        /// </summary>
+        public Guid Id { get; }
+
+        /// <summary>
+        /// Gets the suggestions.
+        /// </summary>
+        public List<string> Suggestions { get; }
+
+        internal PredictionResult(Guid id, List<string> suggestions)
+        {
+            Id = id;
+            Suggestions = suggestions;
+        }
+    }
+
+    /// <summary>
+    /// Provides a set of possible predictions for given input.
+    /// </summary>
+    public static class CommandPrediction
+    {
+        /// <summary>
+        /// Collect the predictive suggestions from registered predictors using the default timeout.
+        /// </summary>
+        /// <param name="ast">The <see cref="Ast"/> object from parsing the current command line input.</param>
+        /// <param name="astTokens">The <see cref="Token"/> objects from parsing the current command line input.</param>
+        /// <returns>A list of <see cref="PredictionResult"/> objects.</returns>
+        public static Task<List<PredictionResult>> PredictInput(Ast ast, Token[] astTokens)
+        {
+            return PredictInput(ast, astTokens, millisecondsTimeout: 20);
+        }
+
+        /// <summary>
+        /// Collect the predictive suggestions from registered predictors using the specified timeout.
+        /// </summary>
+        /// <param name="ast">The <see cref="Ast"/> object from parsing the current command line input.</param>
+        /// <param name="astTokens">The <see cref="Token"/> objects from parsing the current command line input.</param>
+        /// <param name="millisecondsTimeout">The milliseconds to timeout.</param>
+        /// <returns>A list of <see cref="PredictionResult"/> objects.</returns>
+        public static async Task<List<PredictionResult>> PredictInput(Ast ast, Token[] astTokens, int millisecondsTimeout)
```

Comment:
I'm very concerned that an implementer of `IPredictor` will try to access runspace state.  If that happens we may end up with a lot of difficult to troubleshoot dead locks and/or state corruption. This also rules out any possibility of a PowerShell based implementation.

I know y'all are probably looking to work ML into this, but I'm worried this will cause a lot of issues in the long term.

Reply:
Thanks for the feedback!

A key design point is to prevent the predictor implementation from depending on PowerShell to get the prediction results, at least not the default Runspace that is being used by the host.

Time sensitivity is the main reason for this design. The method `PredictInput` is likely to be called for every keystroke, so it needs to finish very fast. I set the default timeout to be 20 ms today, because in testing it seems 30 ms per keystroke is about the most we can have without having a clearly noticeable lagging when typing. The rendering also takes time, so 20 ms timeout seems to be a reasonable default setting for a `PredictInput` call (_it may turn out to be still too much when we hook it up with PSReadLine end-to-end_). With this extreme time limit, `PredictInput` cannot afford running anything in PowerShell.

The second reason is to allow maximum penalization between the render (e.g. PSReadLine) and the predictor. The current implementation of `CommandPrediction.PredictInput` spins up tasks to call `GetSuggestion` on each predictor, and then it returns a task that will finish when either all tasks are done or the timeout is up. From the PSReadLine side, it will call `PredictInput` early (as soon as the user input is known), and continue with its rendering preparation (e.g. `GenerateRender`) until the point that it needs to get the prediction results. That means the a predictor implementation cannot assume that the default Runspace is put on hold when `GetSuggestion` is called, and thus it shouldn't depend on it.

The third reason is about the difference between tab completion and prediction. I don't expect the prediction feature to replace tab completion and it also won't be able to. IMHO, unlike tab completion, prediction doesn't need to be that accurate, so it doesn't need to do things like resolving the `CommandAst` to `CommandProcessor` or `CommandInfo`, which requires the state of the engine where the host is running in.

A predictor implementation should be more ML like, that can learn from history and other telemetry. `CommandPrediction.LineAccepted` is designed to be called when a line is accepted by the host. At that point, the PS engine will be busy running the accepted command line, and the latest history is provided to the predictor to do thinking/learning in parallel. This is a chance for the predictor to spend more time on processing to make up the extreme time limit placed on `Predictor.GetSuggestion`.

Reply:
> A key design point is to prevent the predictor implementation from depending on PowerShell to get the prediction results, at least not the default Runspace that is being used by the host.

Yeah that makes sense.  My worry is that they'll hold onto a bit of state during registration, something that marshals back to the default runspace for a trivial call (like `CommandInfo.Parameters`).

> Time sensitivity is the main reason for this design. The method `PredictInput` is likely to be called for every keystroke, so it needs to finish very fast.

What if there was a one time wait handle, that fires after ~50ms [here](https://github.com/PowerShell/PSReadLine/blob/6cacbebf8d585178c6074191453b6d688a0bd588/PSReadLine/ReadLine.cs#L200).  I'm not sure how the UX would be tbh, it could go either way but I think it's worth trying.  That would greatly limit the amount of the API needs to be called and may end up with a very similar or better experience.

> The third reason is about the difference between tab completion and prediction. I don't expect the prediction feature to replace tab completion and it also won't be able to. IMHO, unlike tab completion, prediction doesn't need to be that accurate, so it doesn't need to do things like resolving the `CommandAst` to `CommandProcessor` or `CommandInfo`, which requires the state of the engine where the host is running in.

There are definitely use cases for it though.  Like finishing a pipeline for instance.  Lets say you have:

```powershell
Get-ChildItem *.log
```

It might be desirable to have a prediction like:

```powershell
Get-ChildItem *.log | Remove-Item
```

Now I'm not saying that's something I personally want, but it does seem inevitable to me that if this API sees a lot of use, there will be some popular state based implementations.

Reply:
@SeeminglyScience Sorry for the delay of my response, and again, I appreciate your feedback!

> What if there was a one time wait handle, that fires after ~50ms here. 

Can you please elaborate a bit more about this? How do you think it will work with a one time wait handle here?


Back to the topic about whether or not allowing predictor to access the default Runspace. Another important reason is that you can have multiple predictors registered, and with the current design, they will be triggered at about the same time. If predictors are allowed to access the default Runspace, then it's likely multiple predictors will try doing that at the same time, which will not work.

Reply:
> Can you please elaborate a bit more about this? How do you think it will work with a one time wait handle here?

One time per key press I mean.  Basically a 50ms timeout after each key press so that it only triggers on a pause or when a prediction is accepted (e.g. right arrow is pressed).

> Back to the topic about whether or not allowing predictor to access the default Runspace. Another important reason is that you can have multiple predictors registered, and with the current design, they will be triggered at about the same time. If predictors are allowed to access the default Runspace, then it's likely multiple predictors will try doing that at the same time, which will not work.

I'm not trying to make a case for directly supporting runspace access, I'm saying if this API is popular then state based implementations are inevitable regardless.  It won't be clear to the implementer exactly how bad that can be, it'll just look like an oversight in the API.

Reply:
We could also take the same approach as what VS Code does. For A-Z characters, they send only one message on the first keystroke and then client side filter as the user types more A-Z characters.

Example:

```
I
```
Gives a bunch of completions like: `Invoke-Command`, `Install-Module`, `Import-Module`... from PSES.

Once the user types:

```
In
```

The _client side filters the results_ from the `I` results and doesn't send a message to PSES. `Import-Module` no longer shows up in completions as expected.

Then `PredictInput` doesn't need to be run as frequently.

Reply:
Thank you both for the helpful ideas! Definitely need to think consider this when hooking up the PSReadLine with the prediction APIs.

Reply:
Feel free to reach out when you get to that point 😃